spec/firebase_app_distribution_action_spec.rb

```diff
@@ -1,141 +1,214 @@
-      end
-
-      it "returns nil if param does not exist" do
-        params = {}
-        expect(action.flag_value_if_supplied('--token', :firebåse_cli_token, params)).to be_nil
+  context '#get_upload_token' do
+    context 'when testing with valid parameters' do
+      it 'should make a GET call to the app endpoint and return the upload token' do
+        expect(temp_connection).to receive(:get).with("/v1alpha/apps/app_id").and_return(fake_get_app_response_valid)
+        upload_token = Fastlane::Actions::FirebaseAppDistributionAction.get_upload_token("app_id", "binary_path")
+        binary_hash = Digest::SHA256.hexdigest(File.open(fake_binary).read)
+        expect(upload_token).to eq(CGI.escape("projects/1/apps/1/releases/-/binaries/#{binary_hash}"))
       end
     end
 
-    describe "flag_if_supplied" do
-      it "returns flag and param value when it exists" do
-        params = {
-            debug: true
-        }
-        expect(action.flag_if_supplied('--debug', :debug, params)).to eq("--debug")
+    context 'when testing with invalid parameters' do
+      it 'should crash if the app has no contact email' do
+        expect(temp_connection).to receive(:get).with("/v1alpha/apps/app_id").and_return(fake_get_app_response_no_email)
+        expect { Fastlane::Actions::FirebaseAppDistributionAction.get_upload_token("app_id", "binary_path") }.to raise_error(ErrorMessage::GET_APP_NO_CONTACT_EMAIL_ERROR)
       end
 
-      it "returns nil if param does not exist" do
-        params = {}
-        expect(action.flag_if_supplied('--debug', :debug, params)).to be_nil
+      it 'should crash if given an invalid app_id' do
+        expect(temp_connection).to receive(:get).with("/v1alpha/apps/invalid_app_id").and_raise(Faraday::ResourceNotFound.new("404"))
+        expect { Fastlane::Actions::FirebaseAppDistributionAction.get_upload_token("invalid_app_id", "binary_path") }.to raise_error(ErrorMessage::INVALID_APP_ID)
       end
-    end
 
-    describe "testers" do
-      it "wraps string parameters as temp files" do
-        params = {
-          testers: "someone@example.com"
-        }
-        expect(fake_file).to receive(:write).with("someone@example.com")
-        expect(action.testers_flag(params)).to eq("--testers-file /tmp/string")
+      it 'should crash if given an invalid binary_path' do
+        expect(File).to receive(:open).with("invalid_binary_path").and_raise(Errno::ENOENT.new("file not found"))
+        expect { Fastlane::Actions::FirebaseAppDistributionAction.get_upload_token("app_id", "invalid_binary_path") }.to raise_error(ErrorMessage::APK_NOT_FOUND)
       end
+    end
+  end
 
-      it "will use a path if supplied" do
-        params = {
-          testers_file: "/tmp/testers_file"
-        }
-        expect(action.testers_flag(params)).to eq("--testers-file /tmp/testers_file")
+  context '#upload_binary' do
```

Comment:
I think this is better as a `describe` because you're describing what this method does below.